### PR TITLE
Local scope

### LINKED ISSUE #50: Implement Local Scope
First version, should be successful for iteration.

### PR DESCRIPTION
**Pull request by alexthemark**
We're getting closer to being a real programming language now. :)
Closes #50 

Implemented as described this morning.

### CODE REVIEW

File: src/words/environment/WordsEnvironment.java [hunk 1]
```diff
@@ -10,13 +12,21 @@
...
 public class WordsEnvironment {
 	private HashMap<String, WordsClass> classes;
-	private HashMap<String, WordsObject> objects;
+	/* gettableObjects is a list of locally scoped symbol tables with the most local 
+	 * scope first in the list, and global scope always being last in the list.
+	 */
+	private LinkedList<HashMap<String, WordsObject>> gettableObjects;
+	/* ungettableObjects contains the objects that have gone out of scope, but should still appear
+	 * on the board, and are still used by action listeners */
+	private HashSet<WordsObject> ungettableObjects;
```
Comment (akallem) [inline src/words/environment/WordsEnvironment.java:21]: Is the advantage of using a HashSet here that, later on when we need to remove an object from all our data structures that reference it, we can do so directly?  As it is now, it seems we don't use the fact that ungettableObjects is a HashSet versus some other structure.

**Comment by akallem** [comment]
Overall this is quite clever and looks very good.

On the tests, could we also have one that tests that an outer scope object is accessible from inner scope (and that this leads to either successful creation or the ObjectAlreadyExists exception)?  I don't see that being tested directly right now.

File: test/words/test/TestINodeRepeat.java [hunk 2]
```diff
@@ -32,4 +33,25 @@ public void testBadRepeat() throws WordsRuntimeException {
...
+	}
+	
+	@Test
+	public void scopeHandlesExceptions() throws WordsRuntimeException {
+		AST statementList = new INodeStatementList(createObjectFred);
+		AST iterativeLoop = new INodeRepeat(fiveLeaf, statementList);
+		loop.enqueueAST(createObjectFred);
+		loop.fastForwardEnvironment(1);
+		// this iterative loop will fail 5 times
+		loop.enqueueAST(iterativeLoop);
```
Comment (akallem) [inline test/words/test/TestINodeRepeat.java:55]: Are you sure this test is working?  It looks to me like you're not fast forwarding the environment to actually run the loop, so the exceptions of trying to create Fred when he already exists in outer scope are never actually generated

**Comment by akallem** [comment]
The one thing I'm not clear on is how scopes are popped when an exception occurs.  See my line comment, I don't think that test is working correctly.

In principle, wouldn't any AST node that caused the environment to enter a new scope need to catch any exceptions that occur within that scope, pop the scope, and then rethrow the exception upwards?

File: src/words/environment/WordsEnvironment.java (hunk 1 above)
Comment (alexthemark) [inline src/words/environment/WordsEnvironment.java:21]: Yep. The whole advantage is when it comes to removal, we have a constant time removal (provided we have a reference to the object).

File: test/words/test/TestINodeRepeat.java (hunk 2 above)
Comment (alexthemark) [inline test/words/test/TestINodeRepeat.java:55]: Now it fast forwards.

**Comment by alexthemark** [comment]
I realized that scopes don't need to be explicitly handled at the exception level (at least for the repeat node). 

The scope is entered and exited on either side of a statement list (INodeRepeat.java, lines 23-25), and each exception is handled at the statement level. If an exception is thrown, it will print to the console from  the statement list evaluated on INodeRepeat.java, line 24. An exception should never bubble up to mess up scope.

File: src/words/environment/WordsEnvironment.java (hunk 1 above)
Comment (akallem) [inline src/words/environment/WordsEnvironment.java:21]: Sweet.

**Comment by akallem** [comment]
Ah, I see what you mean, you're right.  I was thinking any exception would bubble up to the statement list that the repeat statement is a part of, but you're right, it would only bubble up to the statement list contained within the repeat statement, so the scopes are still correct.

**Comment by akallem** [comment]
This looks good to me.

**Pull request merged**